<a href="https://colab.research.google.com/github/umrsid/Cool-LLM-Projects-with-UI/blob/main/chatbot_with_vectorDB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Basic working of Google Palm LLM in LangChain

In [ ]:
!pip install langchain google-generativeai faiss-cpu

In [ ]:
!pip install InstructorEmbedding
!pip install sentence_transformers
import InstructorEmbedding

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 2.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 79.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 96.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 38.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 87.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 81.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 35.4 MB/s eta 0:00:00
  Created wheel for sentence_transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=fae9fded25f29d402de9d8d51681e5a1e081a071936d6c1fc56d4127a1df5856
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence_transformers


/usr/local/lib/python3.10/dist-packages/InstructorEmbedding/instructor.py:7: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import trange


In [ ]:
from langchain.llms import GooglePalm

api_key = 'AIzaSyA59iXeHjrmqnPPAT2TH7_y_SMPRlKc6jA' # get this free api key from https://makersuite.google.com/

llm = GooglePalm(google_api_key=api_key, temperature=0.1)

In [ ]:
poem = llm("Write a 4 line poem of my love for samosa")
print(poem)

**Oh, samosa, my love**
**You are so delicious, so flaky**
**With your filling of potatoes and peas**
**I could eat you every day**


In [ ]:
essay = llm("write email requesting refund for electronic item")
print(essay)

Dear [Name of Customer Service Representative],

I am writing to request a refund for the [product name] that I purchased on [date]. I am not satisfied with the product because [explain why you are not satisfied].

I have attached a copy of my receipt and a picture of the product. I have also included a link to the product's listing on your website.

I would like to request a full refund for the product. I would also like to request that you send me a replacement product.

I have been a loyal customer of your company for many years, and I have never had a problem with a product before. I am disappointed that I am not satisfied with this product, but I hope that you will be able to resolve this issue to my satisfaction.

Thank you for your time and consideration.

Sincerely,
[Your Name]


In [ ]:
from langchain.chains import RetrievalQA


from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from Codebasics FAQ csv file

In [ ]:
!pip install pypdf

from langchain.document_loaders import PyPDFLoader

loader = PyPDFLoader("homo_deus_a_brief_history_of_tomorrow_pdf.pdf")
data = loader.load_and_split()


In [ ]:
data[2]


Document(page_content='In vitro\n fertilisation: mastering creation.\nComputer artwork © KTSDESIGN/Science Photo Library.', metadata={'source': 'homo_deus_a_brief_history_of_tomorrow_pdf.pdf', 'page': 4})

In [ ]:

# loader = CSVLoader(file_path='codebasics_faqs.csv', source_column="prompt")

# # Store the loaded data in the 'data' variable
# data = loader.load()

RuntimeError: ignored

### Hugging Face Embeddings

In [ ]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

e = instructor_embeddings.embed_query("What is your refund policy?")

In [ ]:
len(e)

768

In [ ]:
e[:5]

[-0.0438980795443058,
 0.007685551419854164,
 -0.009231897071003914,
 0.024496253579854965,
 0.03359227254986763]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [ ]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=[data[0]],
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [ ]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

[Document(page_content='Dedication\nTo my teacher, S. N. Goenka (1924–2013),\nwho lovingly taught me important things.', metadata={'source': 'homo_deus_a_brief_history_of_tomorrow_pdf.pdf', 'page': 2})]

As you can see above, the retriever that was created using FAISS and hugging face embedding is now capable of pulling relavant documents from our original CSV file knowledge store. This is very powerful and it will help us further in our project

##### Embeddings can be created using GooglePalm too. Also for vector database you can use chromadb as well as shown below. During our experimentation, we found hugging face embeddings and FAISS to be more appropriate for our use case

In [ ]:
# google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

# from langchain.vectorstores import Chroma
# vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                            persist_directory='./chromadb')
# vectordb.persist()

### Create RetrievalQA chain along with prompt template 🚀

In [ ]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)


### We are all set 👍🏼 Let's ask some questions now

In [ ]:
chain('to whom this book is dedicated')

{'query': 'to whom this book is dedicated',
 'result': 'To my teacher, S. N. Goenka (1924–2013),\nwho lovingly taught me important things.',
 'source_documents': [Document(page_content='Dedication\nTo my teacher, S. N. Goenka (1924–2013),\nwho lovingly taught me important things.', metadata={'source': 'homo_deus_a_brief_history_of_tomorrow_pdf.pdf', 'page': 2})]}

**As you can see above, the answer of question comes from two different FAQs within our csv file and it is able to pull those questions and merge them nicely**

In [ ]:
chain("Do you guys provide internship and also do you offer EMI payments?")

{'query': 'Do you guys provide internship and also do you offer EMI payments?',
 'result': "I don't know.",
 'source_documents': [Document(page_content='Dedication\nTo my teacher, S. N. Goenka (1924–2013),\nwho lovingly taught me important things.', metadata={'source': 'homo_deus_a_brief_history_of_tomorrow_pdf.pdf', 'page': 2})]}

In [ ]:
chain("do you have javascript course?")

In [ ]:
chain("Do you have plans to launch blockchain course in future?")

In [ ]:
chain("should I learn power bi or tableau?")

In [ ]:
chain("I've a MAC computer. Can I use powerbi on it?")

In [ ]:
chain("I don't see power pivot. how can I enable it?")

In [ ]:
chain("What is the price of your machine learning course?")